haddock-api/src/Haddock/Convert.hs

```diff
@@ -610,3 +610,45 @@ synifyFamInst fi opaque = do
     ts' = synifyTypes ts
     annot_ts = zipWith3 annotHsType is_poly_tvs ts ts'
     is_poly_tvs = mkIsPolyTvs (tyConVisibleTyVars fam_tc)
+
+{-
+Note [Invariant: Never expand type synonyms]
+
+In haddock, we never want to expand a type synonym, as we want to keep the
```

Comment:
If you wish to be a little more precise, the goal isn't necessarily to avoid expanding any type synonym ever, but rather to avoid doing so when the type could be presented to the user. This is why this invariant should hold true of all code in `Convert`, since all converted code will eventually be presented to users in Haddocks.

Reply:
Good point. Done.